cpp/src/arrow/testing/random.cc

```diff
@@ -558,5 +584,248 @@ std::shared_ptr<Array> RandomArrayGenerator::ArrayOf(std::shared_ptr<DataType> t
+                                     nan_probability);                                  \
+  }
+
+  const double null_probability =
+      field.nullable()
+          ? GetMetadata<DoubleType>(field.metadata().get(), "null_probability", 0.01)
+          : 0.0;
+  switch (field.type()->id()) {
+    case Type::type::NA:
+      return std::make_shared<NullArray>(length);
+
+    case Type::type::BOOL: {
+      const double true_probability =
+          GetMetadata<DoubleType>(field.metadata().get(), "true_probability", 0.5);
+      return generator->Boolean(length, true_probability, null_probability);
+    }
+
+      GENERATE_INTEGRAL_CASE(UInt8Type);
+      GENERATE_INTEGRAL_CASE(Int8Type);
+      GENERATE_INTEGRAL_CASE(UInt16Type);
+      GENERATE_INTEGRAL_CASE(Int16Type);
+      GENERATE_INTEGRAL_CASE(UInt32Type);
+      GENERATE_INTEGRAL_CASE(Int32Type);
+      GENERATE_INTEGRAL_CASE(UInt64Type);
+      GENERATE_INTEGRAL_CASE(Int64Type);
+      GENERATE_INTEGRAL_CASE_VIEW(Int16Type, HalfFloatType);
+      GENERATE_FLOATING_CASE(FloatType, Float32);
+      GENERATE_FLOATING_CASE(DoubleType, Float64);
+
+    case Type::type::STRING:
+    case Type::type::BINARY: {
+      const int32_t min_length = GetMetadata<Int32Type>(field.metadata().get(), "min", 0);
+      const int32_t max_length =
+          GetMetadata<Int32Type>(field.metadata().get(), "max", 1024);
+      const int32_t unique_values =
+          GetMetadata<Int32Type>(field.metadata().get(), "unique", -1);
+      if (unique_values > 0) {
+        return generator
+            ->StringWithRepeats(length, unique_values, min_length, max_length,
+                                null_probability)
+            ->View(field.type());
+      }
+      return generator->String(length, min_length, max_length, null_probability)
+          ->View(field.type());
+    }
+
+    case Type::type::DECIMAL128:
+    case Type::type::DECIMAL256:
+    case Type::type::FIXED_SIZE_BINARY: {
+      auto byte_width =
+          internal::checked_pointer_cast<FixedSizeBinaryType>(field.type())->byte_width();
+      return generator->FixedSizeBinary(length, byte_width, null_probability)
+          ->View(field.type());
+    }
+
+      GENERATE_INTEGRAL_CASE_VIEW(Int32Type, Date32Type);
+      GENERATE_INTEGRAL_CASE_VIEW(Int64Type, Date64Type);
+      GENERATE_INTEGRAL_CASE_VIEW(Int64Type, TimestampType);
+      GENERATE_INTEGRAL_CASE_VIEW(Int32Type, Time32Type);
+      GENERATE_INTEGRAL_CASE_VIEW(Int64Type, Time64Type);
+      GENERATE_INTEGRAL_CASE_VIEW(Int32Type, MonthIntervalType);
+
+      // This isn't as flexible as it could be, but the array-of-structs layout of this
+      // type means it's not a (useful) composition of other generators
+      GENERATE_INTEGRAL_CASE_VIEW(Int64Type, DayTimeIntervalType);
+
+    case Type::type::LIST: {
+      const int32_t values_length = GetMetadata<Int32Type>(
+          field.metadata().get(), "values", static_cast<int32_t>(length));
```

Comment:
It's slightly odd that list's options aren't more similar to string's: here we specify the total length of the child whereas for strings we specify min/max length.

Reply:
I was mostly mirroring the existing generator functions which presented different APIs, but I'll see if I can't consolidate the cases further for list/string and their large variants. (And maybe then map can delegate to the list case.)